src/lib/Persistence/Legacy/Notification/Gateway/DoctrineDatabase.php

```diff
@@ -82,6 +82,62 @@ public function getNotificationById(int $notificationId): array
         return $query->execute()->fetchAllAssociative();
     }
 
+    /**
+     * @param int[] $notificationIds
+     *
+     * @return int[]
+     */
+    public function bulkUpdateUserNotifications(
+        Notification $notification,
+        bool $pendingOnly = false,
+        array $notificationIds = []
+    ): array {
+        if (!is_int($notification->ownerId) || $notification->ownerId <= 0) {
+            throw new InvalidArgumentException('ownerId', 'Cannot bulk update notifications without valid ownerId');
+        }
+
+        $queryBuilder = $this->connection->createQueryBuilder();
+        $queryBuilder
+            ->select(self::COLUMN_ID)
+            ->from(self::TABLE_NOTIFICATION)
+            ->andWhere($queryBuilder->expr()->eq(self::COLUMN_OWNER_ID, ':ownerId'))
+            ->setParameter(':ownerId', $notification->ownerId, PDO::PARAM_INT);
```

Comment:
Afaik, using `PDO` in this context is incorrect. It may work by accident because DBAL usually doesn't do anything with data (except when array is requested, or some type juggling), but this would definitely not be "stable".

```php
    /**
     * Sets a query parameter for the query being constructed.
     *
     * <code>
     *     $qb = $conn->createQueryBuilder()
     *         ->select('u')
     *         ->from('users', 'u')
     *         ->where('u.id = :user_id')
     *         ->setParameter(':user_id', 1);
     * </code>
     *
     * @param int|string           $key   Parameter position or name
     * @param mixed                $value Parameter value
     * @param int|string|Type|null $type  Parameter type <-------------
     *
     * @return $this This QueryBuilder instance.
     */
    public function setParameter($key, $value, $type = null)
```

```suggestion
            ->setParameter(':ownerId', $notification->ownerId, \Doctrine\DBAL\Types\Type::INTEGER);
```